src/Illuminate/Encryption/EncryptionServiceProvider.php

```diff
@@ -16,35 +16,16 @@ class EncryptionServiceProvider extends ServiceProvider
     public function register()
     {
         $this->app->singleton('encrypter', function ($app) {
-            $config = $app->make('config')->get('app');
+            $config = $app->make('config');
 
-            // If the key starts with "base64:", we will need to decode the key before handing
-            // it off to the encrypter. Keys may be base-64 encoded for presentation and we
-            // want to make sure to convert them back to the raw bytes before encrypting.
-            if (Str::startsWith($key = $this->key($config), 'base64:')) {
-                $key = base64_decode(substr($key, 7));
-            }
+            $currentKey = $config->get('app.encryption.currentKey');
 
-            return new Encrypter($key, $config['cipher']);
-        });
-    }
+            $keyStoreFactory = new KeyStoreFactory();
 
-    /**
-     * Extract the encryption key from the given configuration.
-     *
-     * @param  array  $config
-     * @return string
-     *
-     * @throws \RuntimeException
-     */
-    protected function key(array $config)
-    {
-        return tap($config['key'], function ($key) {
-            if (empty($key)) {
-                throw new RuntimeException(
-                    'No application encryption key has been specified.'
-                );
-            }
+            $keyStore = $keyStoreFactory->make($config);
```

Comment:
The `->make` method takes the entire config? Would it be better if it only took the app.encryption side of the config. Reason why I mention it is because it could encourage unnecessary coupling down the road.